### PR TITLE
feat(benchmark): promote/demote gate on the daily report

### PR DESCRIPTION
**Pull request by jmoreira-valory**
**Stacked on #433** — review that one first; this diff shows only the gate.

#433 makes the report's numbers trustworthy (computed, not LLM-transcribed). This turns them into a decision.

## The gate

| Outcome | Rule |
|---|---|
| **PROMOTE** | `n ≥ 30`, `floor > +0.04`, `condAcc ≥ 50%` |
| **DEMOTE** | `n ≥ 30`, fails that floor, and either `condAcc < 50%` or below its own no-skill baseline **in both windows** |
| **insufficient** | the sample supports neither — stated as such, because "not enough data" ≠ "no improvement" |

Three new columns: **`floor`** (one-sided 95% lower bound on the paired edge — the worst true edge the evidence still supports, and what the promote test actually reads), **`condAcc`** (win rate where the tool disagreed with the price — the only markets a bet is placed on), and **`rec`**.

## Two safety rules override any per-tool verdict

- **Never empty a platform.** When no deployed tool survives, the headline reads `NO ACTION` and the note names a promotable tournament candidate if one exists.
- **A demote must be sustained.** One live tool sat `0.0003` above its floor across 2,055 markets while winning 65% of 1,968 disagreements.

Reliability annotates rather than blocks — a tool that misses calls may still be the best forecaster available, and refusing it can leave a worse one deployed.

## Docs

One page: [`DAILY_REPORT_OPERATOR_GUIDE.md`](benchmark/DAILY_REPORT_OPERATOR_GUIDE.md) — the loop, how to read the report, both gates in a sentence each, which repo to touch, plus an appendix deriving `floor` on two real tools whose deductions differ 22× purely from sample size.

`PROMOTE_DEMOTE_POLICY.md` is **deleted**: its own checkpoint list ran 3 done / 9 pending, both PRs carrying it (#341, #365) closed unmerged, and two of its rules contradicted the shipped code. `PROPOSAL.md` now records at the contradicting line that edge and conditional accuracy are blocking here rather than the soft signals it lists.

## Verification

Every gate branch is mutation-tested — reverting the bound, the condAcc block, the sample floor, the survivor rule, the no-skill margin, the sustained check or the reliability annotation each fails a test that names the case.

### CODE REVIEW

**Review by DIvyaNautiyal07 — COMMENTED** [review]
### P1

**1. `recent is None` silently triggers demote without the required two-window confirmation** — `benchmark/digest_tables.py:~630` in `_verdict_core`:
```python
if _below_no_skill(stats) and (recent is None or _below_no_skill(recent)):
    return "demote: no-skill"
```
The `recent is None` short-circuit means a below-no-skill 90d reading demotes even when `rolling_scores.json` (W-1) is missing entirely. Per the PR body, a demote must be **sustained across both windows** — this treats missing-W-1 as "confirmed bad" rather than "unknown."

Reachable scenarios: first flywheel run on a fresh env, artifact loss on the rolling step, `continue-on-error` skipping the W-1 scoring. All produce a legitimate silent demote of a tool that the policy explicitly says must have two-window confirmation. Fix: `recent is not None and _below_no_skill(recent)`.

### P2

**2. Population SD used where sample SD is expected in the floor formula** — `benchmark/grouping.py`: `variance = sq_sum / edge_n - mean * mean` divides by `edge_n` (population divisor), not `edge_n - 1` (Bessel-corrected sample SD). The floor formula `edge - 1.645 * sd / sqrt(n)` is asymptotically valid with the sample SD; with the population SD the standard error is systematically underestimated by `sqrt((n-1)/n)`. At the minimum `n=30` that's ~1.7% low, so `floor` is slightly too optimistic and can promote a tool the true one-sided 95% bound would refuse. The +0.04 margin absorbs it in practice, and the operator guide acknowledges the approximation — but the code disagrees with what the derivation claims. `_compute_edge_diagnostics` in `scorer.py` uses the same divisor. Divide by `edge_n - 1` in both.

**3. Duplicate constant definitions in `digest_tables.py:~364-376`** — `NO_SKILL_MARGIN = 0.01` appears twice; `Z_ONE_SIDED_95 = 1.645` appears three times. Python's last-wins semantics mean it works today (values agree), but this is a rebase accident and a maintainer changing the constant won't know which occurrence to edit. Delete duplicates.

**4. Missing exact-boundary tests on the three gate thresholds** — `benchmark/tests/test_digest_tables.py`. No test hits `n == 30`, `condAcc == 0.50`, or `floor == 0.04` exactly. Mutations swapping `<` ↔ `<=` or `>` ↔ `>=` at any of these boundaries pass silently. Every test uses safely-away-from-boundary values (n=100 or n=12, edge=0.08 or edge=0.02). Add exact-boundary cases and assert the exact classification.

**5. Z-value mutation is invisible** — `test_floor_column_renders_the_bound` recomputes the expected floor by calling `_edge_lower_bound()` itself. Changing `1.645 → 1.96` in production changes both sides identically; the test still passes. Pin an expected numeric value derived independently (hand calc from known inputs) so a z-value change is caught.

### Nit

**6. 10-line inline prose block inside `_verdict_core`** — `benchmark/digest_tables.py:~598-617`. Paragraph explaining reliability asymmetry and condAcc rationale interrupts the logic flow. Belongs in the function docstring.

**7. `_verdict` return type is bare `str`** — the codebase already uses `Literal[...]` for verdict sets in `roi_sim.py:190` (`RejectReason`, `NoBetReason`) with a comment saying mypy catches a new reason added at one site missing from the closed set. Applying the same pattern here would prevent silent typos in the `_survivors` prefix-match logic.

**8. `_survivors` uses `startswith(("keep", "PROMOTE", "review"))` prefix match** — a future verdict like `"REVIEW_MANUALLY"` matches by accident, and `"Keep"` (case-difference) silently doesn't. A membership check on a Literal-typed set is more robust.

**9. `DAILY_REPORT_OPERATOR_GUIDE.md` floor derivation should move to `PROPOSAL.md`** — the ~30-line statistical derivation (lines 82-111) doesn't help an operator act on a Slack message. `PROPOSAL.md` already owns the rationale. Keep the guide action-focused.

**10. `SCORES_SCHEMA_VERSION` CI check silently ignores import failure** — pre-existing from #433 (`2>/dev/null || echo false` masks scorer import errors). Unlikely to matter (the scorer step would also fail loudly), but worth noting again.

**Review by OjusWiZard — COMMENTED** [review]
Reviewed the gate end to end and reproduced each finding against `1f4be42b`. The gate logic itself is well-built and genuinely well-tested — the promote bound, condAcc, the survivor rule and the blocked headline all have real behavioural tests, and the mutation-testing claim holds up for `digest_tables.py`. Every issue below sits in a **seam**: where a verdict gets re-parsed or truncated by a downstream consumer, or in the accumulator layer underneath.

(I started against `0a8f85d0` — the stale reliability comment is already fixed in `1f4be42b`, so that one's dropped. One neighbouring claim in the same block is still stale; noted inline.)

**Blocking (2)**
- `_verdict_core:505` — `recent` is unvalidated in both directions: an absent W-1 *confirms* a demote, a thin lucky W-1 *erases* one. A scorer outage produces a live `DEMOTE n` recommendation in the same message set that says the week is unmeasured.
- `_headline:309` — the `(rel NN%)` caveat never reaches the headline sentence, and the test that claims to guard it asserts against every flattened message instead.

**Should fix (2)** — both are false statements in an operational Slack message, both small
- `_no_replacement_note:563` — "all N deployed tools demote" counts unjudged tools as demotes (reproduced with this PR's own test fixture).
- `_headline:299` — `"no data"` lands in no headline bucket at all, so a ran-but-unscored tool is invisible.

**Cleanup (3)** — stale condAcc claim at 481, triplicated constants at 249-261, dead `_below_reliability` at 406.

**Your call (2)** — population-vs-sample sd in the floor's standard error, and the untested accumulator layer.

One pattern worth flagging: four of these share a root cause — the verdict is a bare string that three consumers re-parse with `startswith`. An enum plus a single render function would make all four unrepresentable. Follow-up, not a merge blocker.

Verified locally at this head: 1669 tests pass, linters and CI green.

File: benchmark/digest_tables.py [hunk 1]
```diff
@@ -281,12 +362,253 @@ def _headline(platform: str, as_of: str | None = None) -> dict[str, Any]:
...
+            note = f"floor ok, condAcc {conditional:.0%}"
+            return f"keep ({note})" if deployed else f"review: {note}"
+        return "keep (floor ok)" if deployed else "PROMOTE"
+
+    if conditional is not None and conditional < COIN_FLIP:
+        reason = f"condAcc {conditional:.0%}"
+        return f"demote: {reason}" if deployed else f"no: {reason}"
+
+    # Sustained, not a one-off: both the 90d window and the latest week must agree.
+    if _below_no_skill(stats) and (recent is None or _below_no_skill(recent)):
```
Comment (OjusWiZard) [inline benchmark/digest_tables.py]: **Blocking.** `recent` is unvalidated in *both* directions, so the sustained rule fails open and closed at once.

`recent is None` short-circuits the `and` to `True`, and `_below_no_skill(recent)` never applies `MIN_SAMPLE_SIZE` to `recent` — unlike every other cross-window comparison in this file (`_delta`, `_has_floor`, `_alert_rows`). Reproduced at this commit, same 90d stats (500 rows, clearly below no-skill) in all three:

| W-1 | verdict |
|---|---|
| genuinely bad (n=300) | `demote: no-skill` |
| thin & lucky (**n=3**, brier 0.10) | **`keep`** |
| **absent from `w1`** | **`demote: no-skill`** |

Two readings of "we know nothing about last week" give opposite answers, and the *less* informative one takes the destructive action. Conversely a 90-day, 500-sample demote signal is erased by three markets that happened to land well, with nothing on the `keep` cell to say the reprieve rests on n=3.

The absent case is reachable in the run where it hurts most: when the window scorer fails, `windows["w1"]` is `{}` for every tool, and the same message set posts *":warning: W-1 unavailable — those columns are NOT unchanged; they are unmeasured"* directly above a headline reading `DEMOTE n`. It also fires for any tool with no resolutions in 7 days — the `week still filling` state the operator guide tells readers to ignore.

This contradicts the PR body ("A demote must be sustained across both windows"), the operator guide ("in **both** the 90d window and the week"), and the comment on the line above.

One fix covers both directions — an unfloored or absent recent window means "cannot confirm", never "confirmed". `_has_floor(None)` is already `False`, so it subsumes the `is None` case:

```suggestion
    if _below_no_skill(stats) and _has_floor(recent) and _below_no_skill(recent):
```

That also stops the tournament path (`w1=None`) from silently waiving the sustained rule entirely.

File: benchmark/digest_tables.py [hunk 2]
```diff
@@ -243,29 +244,109 @@ def _delta(
...
+        t
+        for t, v in {**prod, **tourn}.items()
+        if v.startswith(("n=", "no spread", "needs --rebuild"))
+    )
+    # Asked of the shared survivor rule, NOT of the verdict text.
+    blocked = bool(prod) and not _survivors(prod)
+
+    if promote:
+        state, token = "promote", f"PROMOTE {len(promote)}"
+        detail = "Promote " + ", ".join(f"`{t}`" for t in promote) + "."
```
Comment (OjusWiZard) [inline benchmark/digest_tables.py]: **Blocking.** The headline silently drops the reliability caveat.

`_verdict` produces `PROMOTE (rel 60%)`, line 297 matches it via `startswith("PROMOTE")`, then this keeps only the tool name. Reproduced at this commit:

```
row verdict:      'PROMOTE (rel 60%)'
headline detail:  'Promote `cand`.'
```

The `(rel NN%)` annotation is the entire mechanism that replaced the hard reliability gate, and it survives only in the `rec` cell of a later message. A reader who acts on the headline promotes a tool scored over a systematically non-random 60% of its calls, having never seen the warning. That is precisely what this function's docstring rules out: *"A reader who stops after the first message must still leave with the right action."*

The test that claims to guard this doesn't. `test_the_headline_announces_an_annotated_promote` asserts `"rel 60%" in _body(...)`, but `_body` flattens **every** message including the tournament table, so it passes on the table row while the headline stays bare — its docstring ("the warning must not mute the headline") is not what it checks.

Carrying the annotated verdict (or just its suffix) into `detail` fixes both.

File: benchmark/digest_tables.py [hunk 3]
```diff
@@ -243,29 +244,109 @@ def _delta(
...
+    :param as_of: last date the data covers, from the scores file -- this
+        module has no clock.
     :return: a webhook payload.
     """
+    promote = sorted(t for t, v in tourn.items() if v.startswith("PROMOTE"))
+    demote = sorted(t for t, v in prod.items() if v.startswith("demote"))
+    unjudged = sorted(
+        t
+        for t, v in {**prod, **tourn}.items()
+        if v.startswith(("n=", "no spread", "needs --rebuild"))
```
Comment (OjusWiZard) [inline benchmark/digest_tables.py]: `"no data"` lands in no headline bucket at all.

`_verdict_core` returns `"no data"` when `brier` is None. That matches neither `"n="`, `"no spread"`, `"needs --rebuild"` here, nor `"PROMOTE"`/`"demote"` above — so the tool is not promoted, not demoted, and not reported as unjudged, while still counting against `_survivors`. `_ran_but_unscored` deliberately puts exactly these tools into `prod_tools`, so this is the normal path for "ran but returned nothing usable".

Combined with the `_no_replacement_note` count bug below, the report contradicts itself on one dataset. Two deployed tools — X ran-but-unscored, Y at n=12 — reproduced end to end:

```
headline:  ⚪ POLYSTRAT · REPORT V2 · NO CHANGE · 2026-08-17
           No candidate clears the promote margin.
           _1 tool(s) had too little data to judge: `Y`._      ← X missing
under 1b:  ⚠️ *NO REPLACEMENT* - all 2 deployed tools demote.   ← neither demotes
```

Stop at the title and you leave with "nothing to do"; scroll and you see a platform-level alarm. Neither is true. No test references `"no data"` anywhere.

File: benchmark/digest_tables.py [hunk 4]
```diff
@@ -281,12 +362,253 @@ def _headline(platform: str, as_of: str | None = None) -> dict[str, Any]:
...
+    ready = sorted(t for t, v in (candidates or {}).items() if v.startswith("PROMOTE"))
+    if ready:
+        names = ", ".join(f"`{t}`" for t in ready)
+        return (
+            f":warning: *NO DEPLOYED REPLACEMENT* - all {len(verdicts)} "
+            f"deployed tools demote, but {names} clears the promote gate in "
+            "the tournament. Promote before retiring, not after."
+        )
+    return (
+        f":warning: *NO REPLACEMENT* - all {len(verdicts)} deployed tools "
```
Comment (OjusWiZard) [inline benchmark/digest_tables.py]: This claims "all N demote" when some tools merely could not be judged.

`_survivors` only excludes `keep`/`PROMOTE`/`review` — it never checks that the remainder actually demoted. Run against *this PR's own test fixture*:

```
in:  {a: "demote: no-skill", b: "demote: no-skill",
      c: "demote: no-skill", d: "n=12 < 30"}
out: ":warning: *NO REPLACEMENT* - all 4 deployed tools demote. ..."
```

Three demote. `test_unjudgeable_tool_is_not_a_replacement` builds exactly this case and asserts only `"NO REPLACEMENT" in note`, so the false count ships green.

Worth noting you fixed this identical class of bug in `_headline` — the `blocked` branch even carries `# NOT len(demote)` — so it's just the sibling that was missed. Same applies to the `NO DEPLOYED REPLACEMENT` string above.

File: benchmark/digest_tables.py [hunk 5]
```diff
@@ -281,12 +362,253 @@ def _headline(platform: str, as_of: str | None = None) -> dict[str, Any]:
...
+        # has no data".
+        if (stats or {}).get("edge_sd", "missing") is None:
+            return "needs --rebuild"
+        return "no spread"
+
+    # A cleared bound is the policy's promote condition and nothing below may
+    # silently override it. Conditional accuracy is a strong diagnostic but it
+    # is NOT in the policy, its 0.50 line carries no interval (today's live
+    # values all straddle it on a Wilson 95%), so it qualifies a promote rather
+    # than vetoing one.
```
Comment (OjusWiZard) [inline benchmark/digest_tables.py]: Thanks for clearing the reliability paragraph in 1f4be42b — this neighbouring claim is stale in the same way, just in the opposite direction.

> `so it qualifies a promote rather than vetoing one`

A candidate with a cleared floor and condAcc < 50% returns `review: ...` a few lines down, and `_headline` promotes only on `startswith("PROMOTE")`. So condAcc **does** veto the promote — as the PR body itself states (`PROMOTE | n ≥ 30, floor > +0.04, condAcc ≥ 50%`).

File: benchmark/digest_tables.py [hunk 6]
```diff
@@ -243,29 +244,109 @@ def _delta(
...
+# is computable from the stored sum/sum-of-squares without per-row data.
+Z_ONE_SIDED_95 = 1.645
+
+NO_SKILL_MARGIN = 0.01
+# One-sided 95%: the normal-approximation z. The policy specifies a bootstrap;
+# at n >= 30 on a mean this is the same call to well inside the margin, and it
+# is computable from the stored sum/sum-of-squares without per-row data.
+Z_ONE_SIDED_95 = 1.645
+
+Z_ONE_SIDED_95 = 1.645
```
Comment (OjusWiZard) [inline benchmark/digest_tables.py]: Rebase artifact: `NO_SKILL_MARGIN` is defined twice (249, 255) and `Z_ONE_SIDED_95` three times (253, 259, 261), with the 3-line comment copy-pasted alongside. Last assignment wins so it's harmless at runtime, and CI is green on it — but it wants a cleanup pass before merge.

File: benchmark/digest_tables.py [hunk 7]
```diff
@@ -281,12 +362,253 @@ def _headline(platform: str, as_of: str | None = None) -> dict[str, Any]:
...
+
+    :param stats: per-tool stats for one window.
+    :return: True when below no-skill by more than the margin.
+    """
+    brier = _num((stats or {}).get("brier"))
+    base = _num((stats or {}).get("baseline_brier"))
+    return brier is not None and base is not None and brier - base >= NO_SKILL_MARGIN
+
+
+def _below_reliability(stats: dict[str, Any] | None) -> bool:
```
Comment (OjusWiZard) [inline benchmark/digest_tables.py:405]: Dead code — added by this PR, never called anywhere in the repo. The same `rate < RELIABILITY_GATE` check is inlined in `_verdict`. Either delete this, or have `_verdict` call it so there's one definition of the gate.

File: benchmark/digest_tables.py [hunk 8]
```diff
@@ -298,7 +620,14 @@ def _cols(names: Iterable[str]) -> tuple[Col, ...]:
     :param names: column header names.
     :return: Col spec tuple for ``table_block``.
     """
-    return tuple(Col(n) if n == "tool" else Col(n, align="right") for n in names)
+    return tuple(
+        (
+            Col(n)
+            if n == "tool"
+            else Col(n, wrap=True) if n == "rec" else Col(n, align="right")
```
Comment (OjusWiZard) [inline benchmark/digest_tables.py]: Nit: the docstring at line 618 still reads *"every other column right-aligned"*, which `rec` now contradicts.

A lookup reads better than the chained ternary and keeps the exception visible:

```python
_COL_KWARGS = {"tool": {}, "rec": {"wrap": True}}
...
return tuple(Col(n, **_COL_KWARGS.get(n, {"align": "right"})) for n in names)
```

File: benchmark/digest_tables.py [hunk 9]
```diff
@@ -641,7 +977,14 @@ def build_digest_messages(
...
 
-    messages: list[dict[str, Any]] = [_headline(platform, as_of)]
+    messages: list[dict[str, Any]] = [
+        _headline(
+            _verdicts_for(sorted(prod_tools), windows["at"], True, windows["w1"]),
+            _verdicts_for(sorted(tourn_tools), windows["tournament"], False),
+            platform,
+            as_of,
+        )
+    ]
```
Comment (OjusWiZard) [inline benchmark/digest_tables.py:989]: Advisory, not for this PR: the gate is recomputed six times per platform per report — twice here, twice inside `_rows_at`, and twice for `_no_replacement_blocks`. Identical today because the call sites happen to pass equivalent tool sets and windows, but nothing binds the headline's verdicts to the table's — and the two message bugs flagged above are both cases of the headline and the table disagreeing about the same data.

Computing `prod_verdicts`/`tourn_verdicts` once here and threading them through would also let `_rows_at` drop its `deployed`/`w1` parameters.

File: benchmark/grouping.py [hunk 10]
```diff
@@ -115,6 +122,25 @@ def _accumulate_group(group: dict[str, Any], row: dict[str, Any]) -> None:
...
+
+    :param group: accumulator dict.
+    :param edge_n: number of edge-eligible rows.
+    :return: standard deviation, or None when it cannot be computed.
+    """
+    sq_sum = group.get("edge_sq_sum")
+    if sq_sum is None or edge_n < 2:
+        return None
+    mean = group["edge_sum"] / edge_n
+    variance = sq_sum / edge_n - mean * mean
```
Comment (OjusWiZard) [inline benchmark/grouping.py]: The floor's standard error uses population sd where sample sd is wanted.

`sq_sum / edge_n - mean * mean` divides by *n*, but `SE = s/√n` in `_edge_lower_bound` wants the Bessel-corrected *s*. At the gate's own n=30 floor with sd=0.3 the floor reads **+0.0015 too high** — about 4% of the 0.04 promote margin — compounding with z=1.645 rather than t₂₉=1.699 (another ~0.003). Both biases push the same way: toward promoting. The operator guide documents the z-vs-t approximation but not this one.

The existing `edge_n < 2` guard already covers the divide:

```suggestion
    variance = (sq_sum - edge_n * mean * mean) / (edge_n - 1)
```

Same applies to the duplicate of this formula at `scorer.py:291`. Since both write the same persisted `edge_sd` field, a shared helper in `scoring_primitives.py` would remove the risk of the two drifting — `test_batch_matches_incremental` omits `edge_sd` from its parity key list, which is exactly where a drift would otherwise surface.

File: benchmark/scorer.py [hunk 11]
```diff
@@ -1033,7 +1040,10 @@ def _finalize_scores(scores: dict[str, Any]) -> dict[str, Any]:
     # re-armed on the incremental path), and stamping it current would stop
     # the flywheel's stale-schema check from ever firing the one rebuild
     # that migrates it.
-    migrated = scores["overall"].get("market_brier_sum") is not None
+    migrated = (
+        scores["overall"].get("market_brier_sum") is not None
+        and scores["overall"].get("edge_sq_sum") is not None
+    )
```
Comment (OjusWiZard) [inline benchmark/scorer.py:1044]: This new `and` clause isn't covered. The only test, `TestSchemaStampEarned.test_pre_migration_accumulator_keeps_its_old_version`, uses a legacy fixture whose `overall` has **neither** key — which the old single-condition check handled identically. The realistic post-schema-2 state (`market_brier_sum` present, `edge_sq_sum` absent) is the whole point of the version bump and is untested; deleting the new clause fails nothing.

Related gaps in the same layer, all deletable without a red test:
- `_edge_sd` (`grouping.py:125`) is never imported by any test — there is no `benchmark/tests/test_grouping.py` at all, so its legacy-`None`, `n < 2` and `variance <= 0` branches are unverified.
- `_restore_group`'s `edge_sq_sum` line (1268) never runs on a tool entry — every resume test passes `by_tool: {}`.
- No persist → resume → incremental-accumulate round-trip for `edge_sq_sum`.

The mutation-testing claim in the PR body holds up well for `digest_tables.py`; it doesn't reach the accumulator the gate depends on. Blast radius is limited — the production window `trailing_scores_*.json` is a fresh nightly batch, so only the tournament accumulator can carry a stale `edge_sq_sum`.

File: benchmark/digest_tables.py (hunk 1 above)
Comment (jmoreira-valory) [reply benchmark/digest_tables.py]: Confirmed and fixed in 324e018d -- and thank you both for the three-row reproduction; it is now literally the test fixture. An absent week and a thin week give the same answer (no confirmation, no demote): the week counts only when it clears MIN_SAMPLE_SIZE like every other cross-window comparison. Candidates keep the single-window non-destructive `no:`. Pinned by `test_unjudgeable_week_never_confirms_a_demote` (absent, thin-lucky, thin-bad all -> keep) and the mutation reverting to `recent is None or ...` fails it.

File: benchmark/digest_tables.py (hunk 2 above)
Comment (jmoreira-valory) [reply benchmark/digest_tables.py]: Confirmed and fixed in 324e018d. The promote detail carries the verdict's suffix: `Promote `cand` (rel 60%).` -- the headline can no longer read cleaner than the rec cell. Pinned by `test_promote_detail_carries_the_rel_annotation`.

File: benchmark/digest_tables.py (hunk 3 above)
Comment (jmoreira-valory) [reply benchmark/digest_tables.py]: Confirmed and fixed in 324e018d: `no data` joins the unjudged bucket, so a ran-but-unscored tool shows in the too-little-data line instead of vanishing. Pinned by `test_no_data_lands_in_the_unjudged_bucket`.

File: benchmark/digest_tables.py (hunk 4 above)
Comment (jmoreira-valory) [reply benchmark/digest_tables.py]: Confirmed and fixed in 324e018d -- the exact sibling of the round-1 `_headline` fix, as you said. Both notes now count demotes: `3 of 4 deployed tools demote and none survives`. Your fixture is the test (`test_mixed_cohort_counts_demotes_not_the_roster`, asserting `3 of 4` present and `all 4` absent), and the weaker existing assertion was strengthened.

File: benchmark/digest_tables.py (hunk 5 above)
Comment (jmoreira-valory) [reply benchmark/digest_tables.py]: Fixed in 324e018d: the comment now says what the code does -- condAcc under 50% VETOES a candidate promote and only annotates a deployed keep.

File: benchmark/digest_tables.py (hunk 6 above)
Comment (jmoreira-valory) [reply benchmark/digest_tables.py]: Squashed in 324e018d -- rebase artifact, one commented set of constants remains.

File: benchmark/digest_tables.py (hunk 7 above)
Comment (jmoreira-valory) [reply benchmark/digest_tables.py:405]: Fixed in 324e018d the way you suggested: `_verdict` now calls `_below_reliability`, making it the single definition of the gate check. (Same lesson as a round-1 #433 finding, where a duplicated inline threshold made a boundary mutation invisible.)

File: benchmark/digest_tables.py (hunk 8 above)
Comment (jmoreira-valory) [reply benchmark/digest_tables.py]: Both taken in 324e018d: docstring corrected and the lookup form adopted.

File: benchmark/digest_tables.py (hunk 9 above)
Comment (jmoreira-valory) [reply benchmark/digest_tables.py:989]: Agreed and deliberately not taken here per your framing. Logged in the loop tracking doc as a follow-up: compute the verdicts once in build_digest_messages and thread them through, which also drops _rows_at's deployed/w1 params. Both round-2 headline/table disagreement bugs are indeed arguments for it.

File: benchmark/grouping.py (hunk 10 above)
Comment (jmoreira-valory) [reply benchmark/grouping.py]: Confirmed and fixed in 324e018d, with your shared-helper suggestion: `sample_edge_sd()` in scoring_primitives is now the one Bessel-corrected definition, used by both the incremental and batch paths (both population-sd duplicates deleted). Branch coverage added (`TestSampleEdgeSd`: known-value, n<2, cancellation clamp), and reverting to the population form fails the known-value test. z-vs-t stays as documented.

File: benchmark/scorer.py (hunk 11 above)
Comment (jmoreira-valory) [reply benchmark/scorer.py:1044]: Covered in 324e018d: `TestSchemaStampPartialMigration` pins the exact state the bump exists for -- `market_brier_sum` armed, `edge_sq_sum` absent -> version stays 2 and the rebuild fires; deleting the new clause fails it. It also asserts the restore keeps `edge_sq_sum` as None rather than the seed. The `sample_edge_sd` branches you flagged as unreachable-by-test are covered in the same commit.

File: benchmark/digest_tables.py [hunk 12]
```diff
@@ -281,24 +361,275 @@ def _headline(platform: str, as_of: str | None = None) -> dict[str, Any]:
...
+    # enough.
+    sustained = _below_no_skill(stats) and (
+        (_has_floor(recent) and _below_no_skill(recent))
+        if deployed
+        else (recent is None or _below_no_skill(recent))
+    )
+    if sustained:
+        return "demote: no-skill" if deployed else "no: no-skill"
+
+    return "keep" if deployed else f"no: {PROMOTE_DELTA - lower:+.3f} short"
```
Comment (OjusWiZard) [inline benchmark/digest_tables.py:512]: Non-blocking, but the one I'd want before this drives real promote/demote traffic: the fallthrough to bare `keep` collapses four materially different states into one cell.

Verified at `324e018d`, same 90d evidence throughout (500 rows, genuinely below no-skill):

| W-1 | verdict |
|---|---|
| agrees, 300 rows | `demote: no-skill` |
| **agrees, 29 rows** (one short of the floor) | **`keep`** |
| **agrees, 25 rows** (week still filling) | **`keep`** |
| **absent** | **`keep`** |
| *disagrees*, 300 rows | `keep` |

Only the last row means "this tool is fine". The middle three mean "could not confirm" — and in two of them the week **agrees with the demote**, so corroborating evidence is discarded and the cell renders identically to a genuine reprieve.

The gate change itself is right, and the conservative direction is the right one (the guide is explicit that a wrong demotion costs more than a missed one). It's the silence I'd flag, not the caution — and it's the same principle this module already enforces everywhere else. `_verdict_core`'s own docstring:

> "not enough data yet" is a different answer from "no improvement", and the policy is explicit that conflating them is the failure to avoid

The promote side has a whole vocabulary for exactly this — `n=… < 30`, `no spread`, `needs --rebuild`, `no data`. The demote side has none. Something like `keep (demote unconfirmed: W-1 n=25 < 30)` would put it on the same cell the reader is already looking at.

A related consequence worth deciding on deliberately: a deployed tool whose weekly window never clears 30 can never demote on the no-skill path at all. The condAcc route still works (it doesn't consult `recent`), so demote isn't dead — I checked, that tool still reads `demote: condAcc 43%` when it loses its disagreements. How often the no-skill route is actually blocked depends on per-tool weekly *resolution* volume, which I couldn't measure from the repo: PROPOSAL's "2k-10k mech requests/day" isn't the right denominator (a row counts only once its market resolves, which is why `week still filling` exists), but your own example tool at 2,055 markets/90d ≈ 160/week clears it comfortably. Your call whether that's a real exposure.

File: benchmark/digest_tables.py [hunk 13]
```diff
@@ -281,24 +361,275 @@ def _headline(platform: str, as_of: str | None = None) -> dict[str, Any]:
...
+
+    # Sustained, not a one-off: both the 90d window and the latest week must agree.
+    # An absent or thin week is "we know nothing new" and must never take
+    # the destructive action; a 3-row lucky week must not erase a 500-row
+    # demote signal either -- the week counts only when it clears the same
+    # sample floor as every other cross-window comparison. Candidates carry
+    # no weekly window; their "no:" is non-destructive, so one reading is
+    # enough.
+    sustained = _below_no_skill(stats) and (
+        (_has_floor(recent) and _below_no_skill(recent))
```
Comment (OjusWiZard) [inline benchmark/digest_tables.py:505]: Second-order effect of depending on `recent` being present *and* floored: when a tool is missing from W-1 entirely, nothing anywhere says so.

`_alert_rows` short-circuits on the tool being present:

```python
starved = [t for t in tools if w1.get(t) and not _has_floor(w1[t])]
```

so an absent tool produces no `sample starved` row. And the round-1 "W-1 unavailable" banner only fires when the **whole** window is empty (`if not windows[key]`), so a *per-tool* absence — a tool with no resolutions that week — falls through the banner, the alert, and the verdict alike. Verified:

```
W-1 thin (n=25)    verdict='keep'   alerts=['sample starved']
W-1 entry ABSENT   verdict='keep'   alerts=[]
```

Three layers of intended loudness, none of which cover this case. Dropping the `w1.get(t) and` guard would at least route it into the starved alert, though putting it on the verdict cell (previous comment) covers both cases at once.

File: benchmark/digest_tables.py [hunk 14]
```diff
@@ -281,24 +361,275 @@ def _headline(platform: str, as_of: str | None = None) -> dict[str, Any]:
...
+            f"{len(verdicts)} deployed tools demote and none survives, but "
+            f"{names} clears the promote gate in the tournament. Promote "
+            "before retiring, not after."
+        )
+    demoted = sum(1 for v in verdicts.values() if v.startswith("demote"))
+    return (
+        f":warning: *NO REPLACEMENT* - {demoted} of {len(verdicts)} deployed "
+        "tools demote and none survives. Acting on these row by row would "
+        "leave the platform with no forecaster at all; treat it as a "
+        "platform-level problem."
```
Comment (OjusWiZard) [inline benchmark/digest_tables.py:573]: The count is honest now, but the sentence around it isn't — this branch is reached whenever *nothing survives*, not only when things demote, so `demoted` can be 0:

```python
_no_replacement_note({"a": "n=12 < 30", "b": "needs --rebuild"})
# ":warning: *NO REPLACEMENT* - 0 of 2 deployed tools demote and none survives.
#  Acting on these row by row would leave the platform with no forecaster at all;
#  treat it as a platform-level problem."
```

"0 of 2 deployed tools demote" under a platform-emergency framing written for a demotion cascade. The actual condition is *nothing is judgeable yet* — a data-availability problem, not a retirement decision, and "acting on these row by row" describes rows that recommend no action. Plausible early in a rollout or during the schema-3 migration window, and no test covers `demoted == 0` (`test_mixed_cohort_counts_demotes_not_the_roster` uses 3 demotes + 1 unjudgeable).

Probably wants a separate sentence when `demoted == 0`. Minor while you're here: `demoted` is computed identically in both branches and can hoist above the `ready` check.
